### PR TITLE
PyTorch Lightning + Heterogeneous Graph Learning Example

### PR DESCRIPTION
**Pull request by rusty1s**

### CODE REVIEW

File: examples/pytorch_lightning/relational_gat.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from torch_geometric import seed_everything
+from torch_geometric.datasets import OGB_MAG
+from torch_geometric.loader import NeighborLoader
+from torch_geometric.nn import GATConv, Linear, to_hetero
+
+
+class DataModule(LightningDataModule):
+    def __init__(self, root: str):
+        super().__init__()
+        self.data = OGB_MAG(root, preprocess='metapath2vec',
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:22]: This should be done in `prepare_data`, as you don't want multi process to execute it and preprocess the data.

File: examples/pytorch_lightning/relational_gat.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        hidden_channels: int,
+        out_channels: int,
+        metadata: Tuple[List[NodeType], List[EdgeType]],
+        dropout: float = 0.5,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+
+        self.model = GAT(hidden_channels, out_channels, dropout)
+        self.model = to_hetero(self.model, metadata, aggr='sum')
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:96]: Add a comment there. Not clear what is happening.

File: examples/pytorch_lightning/relational_gat.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        batch_size = batch['paper'].batch_size
+        y_hat = self(batch.x_dict, batch.edge_index_dict)['paper'][:batch_size]
+        self.val_acc(y_hat.softmax(dim=-1), batch['paper'].y[:batch_size])
+        self.log('val_acc', self.val_acc)
+
+    def test_step(self, batch: Batch, batch_idx: int):
+        batch_size = batch['paper'].batch_size
+        y_hat = self(batch.x_dict, batch.edge_index_dict)['paper'][:batch_size]
+        self.test_acc(y_hat.softmax(dim=-1), batch['paper'].y[:batch_size])
+        self.log('test_acc', self.test_acc)
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:124]: Merge those tests by introducing a common_step.

File: examples/pytorch_lightning/relational_gat.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        self,
+        x_dict: Dict[NodeType, Tensor],
+        edge_index_dict: Dict[EdgeType, Tensor],
+    ) -> Dict[NodeType, Tensor]:
+        return self.model(x_dict, edge_index_dict)
+
+    def training_step(self, batch: Batch, batch_idx: int) -> Tensor:
+        batch_size = batch['paper'].batch_size
+        y_hat = self(batch.x_dict, batch.edge_index_dict)['paper'][:batch_size]
+        loss = F.cross_entropy(y_hat, batch['paper'].y[:batch_size])
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:111]: You are not logging the accuracy for train.

File: examples/pytorch_lightning/relational_gat.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.parameters(), lr=0.01)
+
+
+def main():
+    seed_everything(42)
+    datamodule = DataModule('../../data/OGB')
+    model = RelationalGAT(64, datamodule.num_classes, datamodule.metadata())
+    with torch.no_grad():  # Initialize parameters.
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:134]: We need to find a better way to this :)

File: examples/pytorch_lightning/relational_gat.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        batch_size = batch['paper'].batch_size
+        y_hat = self(batch.x_dict, batch.edge_index_dict)['paper'][:batch_size]
+        y = batch['paper'].y[:batch_size]
+        return y_hat, y
+
+    def training_step(self, batch: Batch, batch_idx: int) -> Tensor:
+        y_hat, y = self.common_step(batch)
+        loss = F.cross_entropy(y_hat, y)
+        self.train_acc(y_hat.softmax(dim=-1), y)
+        self.log('train_acc', self.train_acc, prog_bar=True, on_step=False,
```
Comment (tchaton) [inline examples/pytorch_lightning/relational_gat.py:133]: You might want to log on step during training, especially if there is lot of batches.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
